README.md
# oracle-dba
My collection of Oracle DBA scripts

This repository is a collection of some shell scripts which I wrote over a period of time to help me administer Oracle database servers.

The name of the script files are self-explanatory, so am not spending time in writing detailed description of what these script do. Maybe at a later point in time when I add new scripts which may need explanation, I will add more details here.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- 
### Changed
- 
### Removed
- 

## [0.2] - 2018-05-29
### Added
- Added the CHANGELOG.md file to keep track of changelog
- xe_backup.sh - Fixed the script to cleanup old DMP and log files
- temp_tablespace_usage.sh - Fixed column headings in the output
### Changed
- Renamed expdp_full.sh to xe_backup.sh
- Followed Google shell style guide (https://google.github.io/styleguide/shell.xml)

### Removed
- None

## [0.1] - 2017-06-04
### Added
- expdp_full.sh - Perform full backup of Oracle Database using expdp.
- daily_archive_log.sh - Archivelog generation of an oracle database on a daily basis.
- fra_usage.sh - Display the FRA usage.
- tablespace_free.sh - Displays the space usage of each tablespace in the database.
- temp_tablespace_usage.sh - Current usage of temporary tablespace(sort usage) by active users.


### Changed
- None

### Removed
- None

fra_usage.sh
#!/bin/bash
# purpose: Display the FRA usage
# author:  ajitabhpandey@ajitabhpandey.info
# history: 
#   0.1 on 2017-06-04
#   0.2 on 2018-04-19
#==============================================================================
declare -r TMP_FILE_PREFIX=${TMPDIR:-/tmp}/prog.$$

# prints the usage of the script
function usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# checks the required programs - all programs need to be given as parameters
# e.g. - _check_required_programs expdp find logger
function _check_required_programs() {
  for required_prog in ${@}; do
    hash "${required_prog}" 2>&- || \ 
      { echo >&2 " Required program \"${required_prog}\" not installed or in search PATH.";
        exit 1;
      }
  done
}

function cleanup() {
  rm -f ${TMP_FILE_PREFIX}.*
  echo "Cleaned up temporary file" && exit 100
}

# Sets the ORACLE ENV Variables
function ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [[ -x "$ORA_ENV" ]]; then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -p user.error -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

function main() {
  # set the oracle environment
  ora_env

  sqlplus -s /nolog << __EOF__
  connect / as sysdba
  -- Utilisation (MB) du FRA
  set lines 100
  col name format a60

  SELECT
	  name,
	  floor(space_limit / 1024 / 1024) "Size MB",
	  ceil(space_used / 1024 / 1024) "Used MB"
  FROM V\$RECOVERY_FILE_DEST;

  -- FRA Occupants
  SELECT * FROM V\$FLASH_RECOVERY_AREA_USAGE;

  -- Location and size of the FRA
  SHOW PARAMETER DB_RECOVERY_FILE_DEST;

  -- Size, used, Reclaimable
  SELECT
    ROUND((A.SPACE_LIMIT / 1024 / 1024 / 1024), 2) AS FLASH_IN_GB,
    ROUND((A.SPACE_USED / 1024 / 1024 / 1024), 2) AS FLASH_USED_IN_GB,
    ROUND((A.SPACE_RECLAIMABLE / 1024 / 1024 / 1024), 2) AS
  FLASH_RECLAIMABLE_GB,
    SUM(B.PERCENT_SPACE_USED)  AS PERCENT_OF_SPACE_USED
  FROM
    V\$RECOVERY_FILE_DEST A,
    V\$FLASH_RECOVERY_AREA_USAGE B
  GROUP BY
    SPACE_LIMIT,
    SPACE_USED ,
    SPACE_RECLAIMABLE ;

  -- After that you can resize the FRA with:
  -- ALTER SYSTEM SET db_recovery_file_dest_size=xxG;

  -- Or change the FRA to a new location
  -- (new archives will be created to this new location
  -- ALTER SYSTEM SET DB_RECOVERY_FILE_DEST='/u....';

  exit;
__EOF__

  cleanup

  exit 0
}

# set a trap for cleanup all before process termination by SIGHUBs
trap "cleanup; exit 1" 1 2 3 13 15

# call the main executable function
main "$@"

expdp_full.sh
#!/bin/sh
#script          : expdp_full.sh
#description     : Perform full backup of Oracle Database using expdp.
#author		 : ajitabhpandey@ajitabhpandey.info
#date            : 2017-06-04
#version         : 0.1    
#==============================================================================
#
TIMESTAMP=$(date +%Y%m%d%H)
MYNAME="$(basename $0)"
SCHEMA="FULL"

# prints the usage of the script
usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# Sets the ORACLE ENV Variables
ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [ -x "$ORA_ENV" ]
  then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

while getopts :s:h opt
do
  case ${opt} in
    s  ) SCHEMA=$OPTARG;;
    h  ) usage;;
    \? ) logger -s "Invalid Option: -$OPTARG"
         usage;;
    :  ) logger -s "Invalid Option: -$OPTARG required an argument"
         usage;;
  esac
done
shift $((OPTIND -1))

BKPFILE="$(hostname)_XE_${SCHEMA}_${TIMESTAMP}"

if [ "$SCHEMA" == "FULL" ]
then
  # Take a full backup or oracle database
  expdp \"/ as sysdba\" FULL=Y DIRECTORY=DATA_PUMP_DIR DUMPFILE=${BKPFILE}.DMP LOGFILE=${BKPFILE}.log
else
  # Take backup of specified schema
  expdp \"/ as sysdba\" SCHEMAS=${SCHEMA} DIRECTORY=DATA_PUMP_DIR DUMPFILE=${BKPFILE}.DMP LOGFILE=${BKPFILE}.log
fi

# Find backup location and store it in a variable
BKPLOCATION=$(sqlplus -s /nolog <<EOF
connect / as sysdba
set heading off;
select directory_path "Backup Location" from all_directories
where directory_name='DATA_PUMP_DIR';
exit;
EOF
)

if [ ! -z "$BKPLOCATION" ]
then
  find $BKPLOCATION \( -name "*.DMP" -o -name "*.log" \) -mtime +1 -exec ls -l {} \;
  # Deleting files older than 1 day
  #find $BKPLOCATION -name *.DMP -name *.log -mtime +1 -exec rm -f {} \;
else
  echo "BKPLOCATION variable is not set"
fi



tablespace_free.sh
#!/bin/bash
# purpose: Displays the space usage of each tablespace in the database.
# author: ajitabhpandey@ajitabhpandey.info
# history:
#   0.1 on 2017-06-04
#   0.2 on 2018-04-19
#==============================================================================
declare -r TMP_FILE_PREFIX=${TMPDIR:-/tmp}/prog.$$

# prints the usage of the script
function usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# checks the required programs - all programs need to be given as parameters
# e.g. - _check_required_programs expdp find logger
function _check_required_programs() {
  for required_prog in ${@}; do
    hash "${required_prog}" 2>&- || \ 
      { echo >&2 " Required program \"${required_prog}\" not installed or in search PATH.";
        exit 1;
      }
  done
}

function cleanup() {
  rm -f ${TMP_FILE_PREFIX}.*
  echo "Cleaned up temporary file" && exit 100
}

# Sets the ORACLE ENV Variables
function ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [[ -x "$ORA_ENV" ]]; then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -p user.error -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

function main() {
  # set the oracle environment
  ora_env

  sqlplus -s /nolog << __EOF__
  connect / as sysdba
  set heading on
  set pagesize 500
  set lines 400

  column tablespace       format a30                heading "Tablespace"
  column avail            format 9,999,999,999,999  heading "MB Avail."
  column used             format 9,999,999,999,999  heading "MB Used"
  column free             format 9,999,999,999,999  heading "MB Free"
  column pct              format 999                heading "Pct"

  compute sum of avail used free on report
  break on report
  select  a.tablespace_name "Tablespace",
        a.avail,
        a.avail-b.free used,
        b.free,
        round(nvl((a.avail-b.free)/a.avail*100,0))      "Pct"
  from
  (select tablespace_name, round(sum(bytes)/1048576)     avail
        from    sys.dba_data_files
        group by tablespace_name
        UNION
        select tablespace_name,round(sum(bytes_free+bytes_used)/1048576)
        from v\$temp_space_header
        group by tablespace_name)       a,
  (select tablespace_name, round(sum(bytes)/1048576)     free
        from    sys.dba_free_space
        group by tablespace_name
        UNION
        select tablespace_name,round(sum(bytes_free)/1048576)
        from v\$temp_space_header
        group by tablespace_name)       b
  where  a.tablespace_name = b.tablespace_name (+);

  exit;
__EOF__

  cleanup

  exit 0
}

# set a trap for cleanup all before process termination by SIGHUBs
trap "cleanup; exit 1" 1 2 3 13 15

# call the main executable function
main "$@"

daily_archive_log.sh
#!/bin/bash
# purpose: Archivelog generation of an oracle database on a daily basis.
# author : ajitabhpandey@ajitabhpandey.info
# history : 
#  0.1 on 2017-06-04
#  0.2 on 2018-04-19
#==============================================================================
declare -r TMP_FILE_PREFIX=${TMPDIR:-/tmp}/prog.$$

# prints the usage of the script
function usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# checks the required programs - all programs need to be given as parameters
# e.g. - _check_required_programs expdp find logger
function _check_required_programs() {
  for required_prog in ${@}; do
    hash "${required_prog}" 2>&- || \ 
      { echo >&2 " Required program \"${required_prog}\" not installed or in search PATH.";
        exit 1;
      }
  done
}

function cleanup() {
  rm -f ${TMP_FILE_PREFIX}.*
  echo "Cleaned up temporary file" && exit 100
}

# Sets the ORACLE ENV Variables
function ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [[ -x "$ORA_ENV" ]]; then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -p user.error -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

function main() {
  # set the oracle environment
  ora_env

  sqlplus -s /nolog << __EOF__
    connect / as sysdba
    set pages 1000
    set feedback off
    set verify off
    set echo off
    select trunc(COMPLETION_TIME,'DD') Day, thread#, round(sum(BLOCKS*BLOCK_SIZE)/1048576) MB,count(*) Archives_Generated
    from v\$archived_log
    group by trunc(COMPLETION_TIME,'DD'),thread# order by 1;
    exit;
__EOF__

  cleanup

  exit 0
}

# set a trap for cleanup all before process termination by SIGHUBs
trap "cleanup; exit 1" 1 2 3 13 15

# call the main executable function
main "$@"

temp_tablespace_usage.sh
#!/bin/bash
# purpose: Current usage of temporary tablespace(sort usage) by active users.
# author: ajitabhpandey@ajitabhpandey.info
# history: 
#   0.1 on 2017-06-04
#   0.2 on 2018-04-19
#==============================================================================
declare -r TMP_FILE_PREFIX=${TMPDIR:-/tmp}/prog.$$

# prints the usage of the script
function usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# checks the required programs - all programs need to be given as parameters
# e.g. - _check_required_programs expdp find logger
function _check_required_programs() {
  for required_prog in ${@}; do
    hash "${required_prog}" 2>&- || \ 
      { echo >&2 " Required program \"${required_prog}\" not installed or in search PATH.";
        exit 1;
      }
  done
}

function cleanup() {
  rm -f ${TMP_FILE_PREFIX}.*
  echo "Cleaned up temporary file" && exit 100
}

# Sets the ORACLE ENV Variables
function ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [[ -x "$ORA_ENV" ]]; then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -p user.error -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

function main() {
  # set the oracle environment
  ora_env

  sqlplus -s /nolog << __EOF__
  connect / as sysdba
  set heading on
  set pagesize 500
  set lines 400
  column    sid_serial  format a10
  column    username    format a20
  column    osuser      format a10
  column    spid        format a6
  column    module      format a15
  column    program     format a15
  column    tablespace  format a10
  select S.sid || ',' || S.serial# sid_serial, S.username, S.osuser, P.spid, S.module, S.program, SUM (T.blocks) * TBS.block_size /1024/1024 mb_used, T.tablespace, COUNT(*) sort_ops
  from gv\$sort_usage T, gv\$session S, dba_tablespaces TBS, gv\$process P
  where T.session_addr = S.saddr AND S.paddr = P.addr AND T.tablespace = TBS.tablespace_name
  group by S.sid, S.serial#, S.username, S.osuser, P.spid, S.module, S.program, TBS.block_size, T.tablespace
  order by sid_serial;
  exit;
__EOF__

  cleanup

  exit 0
}

# set a trap for cleanup all before process termination by SIGHUBs
trap "cleanup; exit 1" 1 2 3 13 15

# call the main executable function
main "$@"

xe_backup.sh
#!/bin/sh
# purpose : Perform backup of Oracle Database using expdp.
# author  :	ajitabhpandey@ajitabhpandey.info
# history : 
#  0.1 on 2017-06-04
#  0.2 on 2018-04-18
#==============================================================================
#
declare -r TMP_FILE_PREFIX=${TMPDIR:-/tmp}/prog.$$
declare -r TIMESTAMP=$(date +%Y%m%d%H)
declare -r MYNAME="$(basename $0)"
declare -r SCHEMA="FULL"

# prints the usage of the script
function usage() {
    echo "Usage: " 
    echo "$MYNAME [-s] [-h]"
    exit 1
}

# checks the required programs - all programs need to be given as parameters
# e.g. - _check_required_programs expdp find logger
function _check_required_programs() {
  for required_prog in ${@}; do
    hash "${required_prog}" 2>&- || \ 
      { echo >&2 " Required program \"${required_prog}\" not installed or in search PATH.";
        exit 1;
      }
  done
}

function cleanup() {
  rm -f ${TMP_FILE_PREFIX}.*
  echo "Cleaned up temporary file" && exit 100
}

# Sets the ORACLE ENV Variables
function ora_env() {
  ORA_ENV="/u01/app/oracle/product/11.2.0/xe/bin/oracle_env.sh"
  if [ -x "$ORA_ENV" ]; then
    . ${ORA_ENV}
  else
    echo "Could not find Oracle Environment"
    logger -p user.error -s "$MYNAME - Could not set Oracle Environment"
    exit 1
  fi
}

function main() {
  local -r OPTS=':s:h'

  while builtin getopts ${OPTS} opt; do
    case ${opt} in
      s  ) SCHEMA=$OPTARG
           ;;
      h  ) usage
           ;;
      \? ) logger -p user.error -s "Invalid Option: -$OPTARG"
           usage
           ;;
      :  ) logger -p user.error -s "Invalid Option: -$OPTARG required an argument"
           usage
           ;;
      *  ) logger -p user.error -s "Too many options. You should not see this."
           ;;
    esac
  done
  shift $((OPTIND -1))

  declare -r BKPFILE="$(hostname)_XE_${SCHEMA}_${TIMESTAMP}"

  if [ "$SCHEMA" == "FULL" ]; then
    # Take a full backup or oracle database
    expdp \"/ as sysdba\" FULL=Y DIRECTORY=DATA_PUMP_DIR DUMPFILE=${BKPFILE}.DMP LOGFILE=${BKPFILE}.log
  else
    # Take backup of specified schema
    expdp \"/ as sysdba\" SCHEMAS=${SCHEMA} DIRECTORY=DATA_PUMP_DIR DUMPFILE=${BKPFILE}.DMP LOGFILE=${BKPFILE}.log
  fi

  # Find backup location and store it in a variable
  BKPLOCATION=$(sqlplus -s /nolog <<__EOF__
  connect / as sysdba
  set heading off;
  select directory_path "Backup Location" from all_directories
  where directory_name='DATA_PUMP_DIR';
  exit;
__EOF__
  )

  if [ ! -z "$BKPLOCATION" ]; then
    # Deleting files older than 1 day
    find $BKPLOCATION \( -name "*.DMP" -o -name "*.log" \) -mtime +1 -print -exec rm -f {} \;|logger -p user.info -t dpdump-cleanup
  else
    logger -p user.error -t dpdump-cleanup -s "Cleanup failed - BKPLOCATION variable is not set"

  cleanup

  exit 0
}

# set a trap for cleanup all before process termination by SIGHUBs
trap "cleanup; exit 1" 1 2 3 13 15

# call the main executable function
main "$@"